### PR TITLE
[DA-4019] Updated Re-Pair tool to accept physical measurements and biobank order pairing

### PR DESCRIPTION
**Pull request by DaveK-14**
## Description of changes/additions
Added new logic to the re-pair tool that allows to pair biobank orders and physical measurements, with optional flag for updating finalized site ids as well.

## Tests
- dao_tests/biobank_order_dao_tests.test_repairing_an_order
- dao_tests/physical_measurements_dao_tests.test_repairing_a_record

[DA-4019]: https://precisionmedicineinitiative.atlassian.net/browse/DA-4019?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rdr_service/rdr_client/pairing_assigner.py [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 Usage:
 ./rdr_client/run_client.sh --project all-of-us-rdr-prod --account $USER@pmi-ops.org \
-  pairing_assigner.py file.csv --pairing [site|organization|awardee] \
-  [--dry_run] [--override_site]
+  pairing_assigner.py file.csv --pairing [site|organization|awardee|biospecimen|physical_measurements] \
```
Comment (ayazamlani) [inline rdr_service/rdr_client/pairing_assigner.py]: pm & b are supposed to be optional flags, not finalization. we shouldn't be changing how we pair from site, org or awardee. Only updating the additional fields if --biospecimen flag, --physical_measurements flag, or both. 

Finalization should be updated automatically if these flags are added, no need for an optional flag for finalization.

File: rdr_service/rdr_client/pairing_assigner.py [hunk 2]
```diff
@@ -57,21 +83,44 @@ def main(client):
...
+                    "Malformed participant ID from line %d: %r does not start with P.",
+                    reader.line_num,
+                    participant_id
                 )
                 num_errors += 1
                 continue
 
+            if biospecimen_sync:
+                request_url = "Participant/%s/BiobankOrder" % participant_id
+            elif pm_sync:
```
Comment (ayazamlani) [inline rdr_service/rdr_client/pairing_assigner.py]: please allow for all fields to be populated if both optional flags are added, not one choice or the other.

File: rdr_service/dao/biobank_order_dao.py [hunk 3]
```diff
@@ -824,6 +824,11 @@ def _do_update_with_patch(self, session, order, resource, expected_version):
             order.restoredSiteId = get_site(resource["restoredInfo"])
             order.restoredTime = clock.CLOCK.now()
             order.orderStatus = BiobankOrderStatus.UNSET
+        elif resource["status"].lower() == "re-pairing":
```
Comment (ayazamlani) [inline rdr_service/dao/biobank_order_dao.py:827]: this is cool. I'll need to learn more about the resource['status'] and how it works with re-pairing.

**Review by ayazamlani — CHANGES_REQUESTED** [review]

File: rdr_service/rdr_client/pairing_assigner.py [hunk 4]
```diff
@@ -27,17 +27,25 @@
...
+        "processedInfo",
+        "finalizedInfo"
+    ]
+
     pairing_key = client.args.pairing
 
-    if client.args.pairing not in pairing_list:
-        sys.exit("Pairing must be one of site|organization|awardee")
+    if client.args.pairing not in p_pair_list and client.args.pairing not in ps_pair_list:
+        sys.exit("Pairing must be one of site|organization|awardee|biospecimen|physical_measurements")
```
Comment (ayazamlani) [inline rdr_service/rdr_client/pairing_assigner.py]: no need for biospecimen|physical_measurements in this sys.exit()

File: rdr_service/rdr_client/pairing_assigner.py [hunk 5]
```diff
@@ -132,7 +195,9 @@ def _get_old_pairing(participant, pairing_key):
     arg_parser.add_argument("file", help="The name of file containing the list of HPOs and participant IDs")
     arg_parser.add_argument("--dry_run", action="store_true")
     arg_parser.add_argument(
-        "--pairing", help="set level of pairing as one of" "[site|organization|awardee]", required=True
+        "--pairing",
+        help="set level of pairing as one of" "[site|organization|awardee|biospecimen|physical_measurements]",
```
Comment (ayazamlani) [inline rdr_service/rdr_client/pairing_assigner.py]: also pls remove biospecimen|physical_measurements here

**Review by ayazamlani — CHANGES_REQUESTED** [review]
there are some auto reformatting happening that is getting pushed up making it difficult to decipher which code is yours. Please try to only push up your code changes so we can easily see what you changed.

File: tests/dao_tests/test_biobank_order_dao.py [hunk 6]
```diff
@@ -395,3 +395,20 @@ def test_ignored_order(self):
...
+        order = self._make_biobank_order()
+        order_json = self.dao.to_client_json(order)
+        order.sourceSiteId = 2
+        order.collectedSiteId = 2
+        order.processedSiteId = 2
+        order.finalizedSiteId = 1
+        order_json['status'] = 're-pairing'
+        self.dao.insert(order)
+        updated_order = self.dao.update_with_patch(order.biobankOrderId, order_json, order.version)
+        self.assertNotEqual(updated_order.sourceSiteId, order.sourceSiteId)
```
Comment (ayazamlani) [inline tests/dao_tests/test_biobank_order_dao.py]: can we see the new site ID number in the test and assertEquals that the updated_order .* 
 = new_site_id

File: tests/dao_tests/test_physical_measurements_dao.py [hunk 7]
```diff
@@ -461,3 +461,20 @@ def test_core_data_qualifiers(self):
...
+        measurement.finalizedSiteId = 1
+        measurement_json['status'] = 're-pairing'
+        self.dao.insert(measurement)
+        with self.dao.session() as session:
+            updated_measurement = self.dao.update_with_patch(
+                measurement.physicalMeasurementsId,
+                session,
+                measurement_json
+            )
+        self.assertNotEqual(measurement.createdSiteId, updated_measurement.createdSiteId)
```
Comment (ayazamlani) [inline tests/dao_tests/test_physical_measurements_dao.py:485]: same here, can we pls see the new site ID and assertEquals the new_site_id

**Review by ayazamlani — CHANGES_REQUESTED** [review]

File: tests/dao_tests/test_physical_measurements_dao.py [hunk 8]
```diff
@@ -461,3 +461,29 @@ def test_core_data_qualifiers(self):
         ).one()
         self.assertTrue(db_obj.satisfiesHeightRequirements)
         self.assertTrue(db_obj.satisfiesWeightRequirements)
+
+    def test_repairing_a_record(self):
```
Comment (dtharpeuno) [inline tests/dao_tests/test_physical_measurements_dao.py:465]: :nit should be "an"

**Review by dtharpeuno — APPROVED** [review]

**Review by ayazamlani — APPROVED** [review]

**Pull request merged**